src/stdlib_math_linspace.fypp

```diff
@@ -0,0 +1,78 @@
+#:include "common.fypp"
+#:set RANKS = range(1, MAXRANK + 1)
+#:set RC_KINDS_TYPES = REAL_KINDS_TYPES + CMPLX_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_linspace
+
+implicit none
+
+contains
+
+  #:for k1, t1 in REAL_KINDS_TYPES
+    #:set RName = rname("linspace_default", 1, t1, k1)
+    module function ${RName}$(start, end) result(res)
+      ${t1}$, intent(in) :: start
+      ${t1}$, intent(in) :: end  
+        
+      ${t1}$ :: res(DEFAULT_LINSPACE_LENGTH)
+
+      res = linspace(start, end, DEFAULT_LINSPACE_LENGTH)
+
+    end function ${RName}$
+  #:endfor
+
+  #:for k1, t1 in REAL_KINDS_TYPES
+    #:set RName = rname("linspace_n", 1, t1, k1)
+    module function ${RName}$(start, end, n) result(res)
+      ${t1}$, intent(in) :: start
+      ${t1}$, intent(in) :: end  
+      integer, intent(in) :: n
+        
+      ${t1}$ :: res(n)
+
+      integer :: i    ! Looping index
+      ${t1}$ :: interval ! Difference between adjacent elements
+
+
+      if(n <= 0) return ! If passed length is less than or equal to 0, return an empty (allocated with length 0) array
+      if(n == 1) then
+        res(1) = end
+        return
+      end if
+
+      interval = (end - start) / real((n - 1), ${k1}$)
+
+      do i = 1, n
```

Comment:
I think this can safely be a `do concurrent`. What do you think?

Reply:
I have been bitten several times with `do concurrent` now, especially for concurrent but not parallel cases, which were parallelized by certain compilers regardless and unsurprisingly broke. The best strategy here is a lot of testing, but we will currently miss the quite aggressive parallelization of concurrencies by `ifort -qopenmp` in our CI setup because we are not compiling stdlib with OpenMP support. I'll open a separate issue to discuss this.

The above one seems safe to do as parallel concurrency.

Reply:
I wonder if using do concurrent here would be setting a wrong precedent. There are many situations where the loop iterations are independent but I would not want to see parallel execution. Typically the largest benefits are delivered if `do concurrent` is used at the highest possible level of the problem. `linspace` is more of a helper routine used to prepare for the bulk problem (which is to be solved with concurrently).

We might also consider writing this loops as follows:
```
res(1) = start
do i = 2, n-1
   res(i) = res(i-1) + interval
end do
res(n) = end  ! make sure end matches user provided one
```

Reply:
I mainly agree with @ivan-pi . I would typically use such a procedure in a parallel section.Therefore, I would expect that such a procedure is sequential (like for intrinsics such as `sum`).

Reply:
In this case and in many others it's inconsequential. A good compiler can auto-parallelize regular `do`-loops. `do concurrent` simply provides more information, and is useful as code annotation as well. I've seen occasional poorer performance of `do concurrent` over regular loops, and I've seen occasional better performance. I have never seen a `do concurrent` break a program when used for independent iterations.

Since it's controversial, best to not do `do concurrent` here until it gets more support from the community.

Reply:
@ivan-pi made an important suggestion that I think may have gotten overlooked in this discussion about `do concurrent`. As the do-loop is currently written (or a `do concurrent` equivalent), the final value `res(n)` will not generally be exactly the same as `end` due to round off error. I think the legitimate expectation of a user is that the first and last values are bit-identical to the input `end` and `start` values.

Reply:
In that case I will modify the code such that the do loop executes from 2 to `n-1`, such that `res(1)` and `res(n)` will be set equal to, and therefore be bit-identical, to the `end` and `start` values.

Reply:
Since a different issue has been opened discussing `do concurrent` and given that I've implemented bit-identical implementations and checks in `test_logspace.f90`, I am marking this conversation as resolved.